### DISCUSSION TITLE
PCA denoising patch_size requirement

### DISCUSSION THREAD

**Discussion by samcoveney** [dipy/dipy#2578] [Reconstruction]
The routine `genpca()` in localpca.py has the requirement that the number of pixels used in a patch exceeds the number of acquisitions. See this code from localpca.py

```
if np.prod(patch_size) < arr.shape[-1]:
    e_s = "You asked for PCA denoising with a "
    e_s += "patch_radius of {0} ".format(patch_radius)
    e_s += "with total patch size of {0} ".format(np.prod(patch_size))
    e_s += "for data with {0} directions. ".format(arr.shape[-1])
    e_s += "This would result in an ill-conditioned PCA matrix. "
    e_s += "Please increase the patch_radius."
    raise ValueError(e_s)
```

The number of pixels is the number of data points N, with the number of acquisitions being the number of dimensions K (see [here](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0073021)). In other words, N points in K-dimensional space.

It is usually preferable to have more samples than dimensions, but this is not a requirement for PCA. Imagine the case where we had many repeated acquisitions with different bvecs, so much so that we had K=10000. We would expect to be able to represent this data with fewer eigenvectors than 10000, yet DIPY would insist that we choose a patch_size with at least 10000 voxels. This seems unreasonable to me.

To put it another way, let's take the case in [this paper](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0073021) with 4x4x4 patches i.e. N = 64, and K = 60 directions. Here N > K so DIPY will work fine. But if the authors were to suddenly discover that they threw away 75% of their data due to a coding error, and that they really have 240 acquisitions (4 repeated measurements per direction for 60 directions), then DIPY would now insist that N = 64 is not big enough (even though performing PCA with a patch of 64 pixels for 240 acquisitions might show that nearly all of the signal was found in the first 40 components).

Any thoughts on this?

**Comment by ShreyasFadnavis** [comment]
Hi @samcoveney -- this is a good point! I think a workaround would be to default to SVD (since it can deal with ill-posed matrices too) if N>K with a warning telling the users that PCA will still function but may not get you the desired output every time. 

PS - This is one of the issues that Patch2Self fixes: removing the need to tune patch sizes, etc.

**Comment by samcoveney** [comment]
Hi @ShreyasFadnavis thanks for your reply.

I am working on a fork to allow `genpca()` to work on data where some spatial dimensions are size 1.

I can include the change to allow for using N < K -- are you suggesting to raise an explicit warning to the user for these cases, or just to include such warnings in the documentation? If working with 2D data, we are much more likely to run into this issue, and I don't know whether to trouble the user with these warnings every time. I could include a function argument to turn off such warnings (default False) but that seems like it might be equally irritating.

Thanks. More thoughts on this matter would be appreciated.

**Reply by samcoveney** [reply]
See PR #2579
